### PR TITLE
Pull test run data from endpoint

### PR DESCRIPTION
**Pull request by kntsoriano**
## Description
Since the test run data can be really big, we need to pull it from a separate endpoint. This PR does that.

### CODE REVIEW

File: app/controllers/api/submissions/test_runs_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module API
+  module Submissions
+    class TestRunsController < BaseController
+      def show
+        submission = current_user.submissions.find_by!(uuid: params[:submission_id])
+
+        test_run = if submission.tests_queued?
```
Comment (kntsoriano) [inline app/controllers/api/submissions/test_runs_controller.rb]: @iHiD Is it fine to place this here? Or should we add it to the serializer?

Comment (iHiD) [reply app/controllers/api/submissions/test_runs_controller.rb]: I don't think we should mock out a fake test-run. I think if the test run doesn't exist, we should be returning a `404` with `json: {test_run: nil}` and then the React should deal with that.

I presume the React can already cope with that anyway, as it's the same as the initial state of the exercise? And the React should also be rendering something for a missing test-run than an empty test-run I'd have thought?

File: app/javascript/components/student/editor/TestRunSummary.tsx [hunk 2]
```diff
@@ -3,16 +3,16 @@ import { Submission, TestRun, TestRunStatus } from '../Editor'
 import { TestRunChannel } from '../../../channels/testRunChannel'
 import { TestRunSummaryContent } from './TestRunSummaryContent'
 import { fetchJSON } from '../../../utils/fetch-json'
+import { typecheck } from '../../../utils/typecheck'
+import { camelizeKeys } from 'humps'
 
 export function TestRunSummary({
-  testRun,
+  submission,
```
Comment (iHiD) [inline app/javascript/components/student/editor/TestRunSummary.tsx]: I'm not clear on why you've changed this in this file. It feels like we've more tightly coupled things, rather than losening the coupling. Can you explain?

File: test/system/components/student/editor_test.rb [hunk 3]
```diff
@@ -1,106 +1,119 @@
...
+          wait_for_submission
+          click_on "Cancel"
 
-        assert_text "Status: cancelled"
+          assert_text "Status: cancelled"
+        end
       end
 
       private
       def wait_for_submission
```
Comment (ErikSchierboom) [inline test/system/components/student/editor_test.rb:115]: This method name implies that it is waiting for something. With the removal of the `wait` it no longer waits. Maybe inline this method, or else add a `wait_for_cancellation` method to make the 

```
click_on "Run tests"
wait_for_submission

click_on "Cancel"
assert_text "Status: cancelled"
```

be slightly more consistent.

File: app/javascript/components/student/editor/TestRunSummary.tsx (hunk 2 above)
Comment (ErikSchierboom) [reply app/javascript/components/student/editor/TestRunSummary.tsx]: It looks like the reason for this change is the use in several places of `submission.links` and `submission.uuid`. So instead of passing in the test run and the submission properties separately, it provides just the submission. Is that correct?

**Review by ErikSchierboom — APPROVED** [review]

File: app/javascript/components/student/editor/TestRunSummary.tsx (hunk 2 above)
Comment (kntsoriano) [reply app/javascript/components/student/editor/TestRunSummary.tsx]: Yes. It depends on a lot of things on submission that's why I just passed submission down.

File: app/controllers/api/submissions/test_runs_controller.rb (hunk 1 above)
Comment (kntsoriano) [reply app/controllers/api/submissions/test_runs_controller.rb]: Cool. I will do this in React then. I was just thinking about having some kind of consistency with a "null object" which the server supplies, instead of the client just guessing what it is. It's fine though.

File: test/system/components/student/editor_test.rb (hunk 3 above)
Comment (kntsoriano) [reply test/system/components/student/editor_test.rb:115]: Will do :)

File: app/controllers/api/submissions/test_runs_controller.rb (hunk 1 above)
Comment (SleeplessByte) [reply app/controllers/api/submissions/test_runs_controller.rb]: I _agree_ with `404`, but i'd return:

```ruby
render json: { errors: [ { message: 'Test run doesn\'t exist.' } ] }, 
           status: :not_found, 
           content_type: 'application/vnd.exercism.errors.v1+json'
```

On the _react_ side, this is a regular 404, with a *structured* error message. You can "just" use `response.json()`, like regular, but you know for a fact that it will be `errors: [ array of errors ]`.

It's probably a _good_ idea, to use `rescue_from ActiveRecord::RecordNotFound` in the `api` base controller, and rescue from with that response. You *can* use a simple gem [`erratum_responsum`](https://github.com/XPBytes/erratum_responsum), which adds all the regular api error codes as messages. This is what we use:

```ruby
  # initializers/<...>.rb  
  self.error_media_type = 'application/vnd.xpbytes.errors.v1+json'

  # api_controller.rb
  include ErratumResponsum

  rescue_from CanCan::AccessDenied, AuthorizedTransaction::TransactionUnauthorized, with: :forbidden
  rescue_from OptimisticallyStale::MissingLockVersion, ActionController::BadRequest,
              RequestMissingParam, with: :bad_request
  rescue_from ActiveRecord::RecordNotFound, with: :not_found
  rescue_from ActionController::UnknownFormat, NoAcceptSerializer, with: :not_acceptable
  rescue_from ResourceGone, with: :gone
  rescue_from ActiveRecord::StatementInvalid, ActiveRecord::RecordInvalid,
              ContentDoesNotMatchContentType, with: :unprocessable_entity
  rescue_from ActiveRecord::RecordNotUnique, ActiveRecord::StaleObjectError, with: :conflict
  rescue_from ContentTypeNotRecognised, with: :unsupported_media_type
```

Additionally, it gives you [_stable_ error codes](https://github.com/XPBytes/erratum_responsum/blob/de6590825bb74625058460819c3944dc4f2cb212/lib/erratum_responsum.rb#L82-L91) you can track:
```ruby
  # inside the gem
  
  def serialize_errors(errors)
    Array(errors).map do |error|
      { message: error.respond_to?(:message) ? error.message : error, code: error_code(error) }
    end
  end

  def error_code(error)
    return 'Ex' + String(error.send(:error_code)) if error.respond_to?(:error_code)
    'Gx' + Digest::MD5.hexdigest(error.class.name)
  end
```

File: app/javascript/components/student/Editor.tsx [hunk 4]
```diff
@@ -58,6 +59,7 @@ export enum TestStatus {
 type SubmissionLinks = {
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:60]: At some point I recommend to switch to

```typescript
{
  cancel: { href: string },
  submit: { href: string },
  testRun: { href: string }
}
```

It allows for metadata and parameters to be part of the URL value, and doesn't require adding weird types later to do this.

File: app/javascript/components/student/Editor.tsx [hunk 5]
```diff
@@ -11,12 +11,13 @@ import { FileEditor, FileEditorHandle } from './editor/FileEditor'
...
 
 export type Submission = {
   testsStatus: SubmissionTestsStatus
   uuid: string
   links: SubmissionLinks
-  testRun: TestRun
+  testRun?: TestRun
 }
 
 type APIError = {
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:24]: I recommend using the error type given above. There will be times where compound errors will be a thing:
```json
{
  "errors": [
    { "message": "Something went wrong" },
    { "message": "Second error", "param": "value", "code": "whatever" }
  ]
}
```

Nesting inside `errors` allows _request_ metadata to be present at the top-level.

File: app/javascript/components/student/Editor.tsx [hunk 6]
```diff
@@ -11,12 +11,13 @@ import { FileEditor, FileEditorHandle } from './editor/FileEditor'
 import { fetchJSON } from '../../utils/fetch-json'
 import { typecheck } from '../../utils/typecheck'
 import { Iteration } from '../track/IterationSummary'
+import { useIsMounted } from 'use-is-mounted'
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:14]: ✔️

File: app/javascript/components/student/Editor.tsx [hunk 7]
```diff
@@ -11,12 +11,13 @@ import { FileEditor, FileEditorHandle } from './editor/FileEditor'
...
 
 export type Submission = {
   testsStatus: SubmissionTestsStatus
   uuid: string
   links: SubmissionLinks
-  testRun: TestRun
+  testRun?: TestRun
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:21]: FYI, this does *not* means `testRun` can be `TestRun` or `undefined`. It means: `testRun` may not be present. Doesn't matter here, but might at some point in the future.

File: app/javascript/components/student/Editor.tsx [hunk 8]
```diff
@@ -11,12 +11,13 @@ import { FileEditor, FileEditorHandle } from './editor/FileEditor'
 import { fetchJSON } from '../../utils/fetch-json'
 import { typecheck } from '../../utils/typecheck'
 import { Iteration } from '../track/IterationSummary'
+import { useIsMounted } from 'use-is-mounted'
 
 export type Submission = {
   testsStatus: SubmissionTestsStatus
   uuid: string
```
Comment (SleeplessByte) [inline app/javascript/components/student/Editor.tsx:19]: Instead of passing back the `uuid`, add ` link` to the `links` object that refers to this, e.g.
```typescript
{ self: { href: 'string' } }
```

Reasons:

- discourages building URLs.
- allows testsubmissions to live on different domains (can track this later).
- don't need an extra URL later if you want to re-request it.

You might ask: "But we already know the test submission URL, right?"

Yes, but it might redirect. This would be the canonical URL :)

File: app/javascript/components/student/editor/TestRunSummary.tsx [hunk 9]
```diff
@@ -78,14 +82,22 @@ export function TestRunSummary({
...
+  useEffect(() => {
+    channel.current = new TestRunChannel(submission, (testRun: TestRun) => {
       setTestRun(testRun)
     })
 
     return () => {
       channel.current?.disconnect()
     }
-  }, [testRun.submissionUuid])
+  }, [submission.uuid])
```
Comment (SleeplessByte) [inline app/javascript/components/student/editor/TestRunSummary.tsx]: Here you would use the link.self.href instead of the uuid.

File: app/serializers/serialize_submission.rb [hunk 10]
```diff
@@ -15,9 +15,12 @@ def call
           submission.solution.uuid,
           submission_id: submission.uuid,
           auth_token: user.auth_tokens.first.to_s
+        ),
+        test_run: Exercism::Routes.api_submission_test_run_url(
+          submission.uuid,
```
Comment (SleeplessByte) [inline app/serializers/serialize_submission.rb:20]: This is fine, because this is at serialization. Here you can build the URL and then never again.

**Review by SleeplessByte — COMMENTED** [review]
A few pointers that will take *less* time to implement now, and make your life easier in the future. I *strongly* recommend doing these now.

File: app/controllers/api/submissions/test_runs_controller.rb (hunk 1 above)
Comment (kntsoriano) [reply app/controllers/api/submissions/test_runs_controller.rb]: This is nice! I'll put this in a separate issue.

File: test/system/components/student/editor_test.rb (hunk 3 above)
Comment (kntsoriano) [reply test/system/components/student/editor_test.rb:115]: I thought about this again, and I think it's fine. When capybara asserts things, it waits for the result implicitly. I wouldn't also want to inline this as it takes the form of `assert_text` which sort of means like it's an assertion. But this is the only way for us to ensure that we wait for the API request to get back.

File: app/javascript/components/student/Editor.tsx (hunk 8 above)
Comment (kntsoriano) [reply app/javascript/components/student/Editor.tsx:19]: In our websockets connection, we would need the UUID. How do you think should we deal with this?

File: app/javascript/components/student/Editor.tsx (hunk 4 above)
Comment (kntsoriano) [reply app/javascript/components/student/Editor.tsx:60]: Cool. I'll put this in a separate issue.

File: app/javascript/components/student/editor/TestRunSummary.tsx (hunk 2 above)
Comment (kntsoriano) [reply app/javascript/components/student/editor/TestRunSummary.tsx]: I've updated the PR and I've moved things back to how they were before. Agree with making the coupling less loose.

**Comment by iHiD** [comment]
(@kntsoriano I'm on Slack btw if you're around :))

File: app/javascript/components/student/Editor.tsx [hunk 11]
```diff
@@ -269,6 +279,18 @@ export function Editor({
     return abort
   }, [abort])
 
+  useEffect(() => {
+    if (!submission) {
+      return
+    }
+
+    fetchJSON(submission.links.testRun, {
```
Comment (ErikSchierboom) [inline app/javascript/components/student/Editor.tsx:319]: Do we need to add the abort code used in https://github.com/exercism/v3-website/pull/249#pullrequestreview-524090289 here too?

File: app/javascript/components/student/editor/TestRunSummary.tsx [hunk 12]
```diff
@@ -82,22 +78,14 @@ export function TestRunSummary({
...
-    fetchJSON(submission.links.testRun, {
-      method: 'GET',
-    }).then((json: any) => {
-      setTestRun(typecheck<TestRun>(camelizeKeys(json), 'testRun'))
-    })
-  }, [submission.uuid])
-
-  useEffect(() => {
-    channel.current = new TestRunChannel(submission, (testRun: TestRun) => {
+    channel.current = new TestRunChannel(testRun, (testRun: TestRun) => {
```
Comment (ErikSchierboom) [inline app/javascript/components/student/editor/TestRunSummary.tsx]: So much cleaner!

Comment (ErikSchierboom) [reply app/javascript/components/student/editor/TestRunSummary.tsx]: I do find the re-using of the `testRun` name for both the argument and the lambda parameter slightly confusing. Maybe you could use a different name for the lambda parameter?

**Review by ErikSchierboom — APPROVED** [review]

File: app/javascript/components/student/Editor.tsx (hunk 11 above)
Comment (kntsoriano) [reply app/javascript/components/student/Editor.tsx:319]: Yes. I've added this now :)

File: app/javascript/components/student/editor/TestRunSummary.tsx (hunk 12 above)
Comment (kntsoriano) [reply app/javascript/components/student/editor/TestRunSummary.tsx]: Agreed. I renamed it to `updatedTestRun`.

**Review by iHiD — APPROVED** [review]
I'm happy to merge this from the Ruby side of things btw :)

File: app/javascript/components/student/Editor.tsx (hunk 8 above)
Comment (kntsoriano) [reply app/javascript/components/student/Editor.tsx:19]: @SleeplessByte Merging this one now, but feel free to reply to this one.

Comment (SleeplessByte) [reply app/javascript/components/student/Editor.tsx:19]: Do you need the uuid, or the _channel_?

It might be that the channel is the uuid, but we don't want to call it uuid then. Instead, subscription channel or whatever.

File: test/system/components/student/editor_test.rb (hunk 3 above)
Comment (SleeplessByte) [reply test/system/components/student/editor_test.rb:115]: That's a perfectly fine reason. Should probably document it inside at `wait_for_submission`, to explain WHY.

**Comment by iHiD** [comment]
Thanks @kntsoriano and @SleeplessByte 💙
